cashctrl_ledger/ledger.py

```diff
@@ -1199,6 +1199,76 @@ def _add_fx_adjustment(
         else:
             raise ValueError("Expecting at least one `entry` row.")
 
+    def FX_revaluation(self, accounts: pd.DataFrame = None, date: datetime.date = None):
+        """Record foreign exchange gains or losses.
+
+        Revalue given foreign currency accounts based on the provided exchange
+        rates (if not provided, use all foreign currency accounts)
+        and record the revaluation gain/losses.
+
+        Args:
+            accounts (pd.DataFrame, optional): DataFrame with the following columns:
+                - 'foreign_currency_account' (int): The account to revalue.
+                - 'exchange_rate' (float): The exchange rate to apply.
+                - 'fx_gain_loss_account' (int): The account in which to record FX gain/loss.
+            date (datetime.date, optional): The date for the revaluation.
+                Defaults to the last day of the current accounting year.
+        """
+
+        def get_fx_gain_loss_account() -> int:
+            """Retrieves the FX gain/loss account from the settings."""
+            settings = self._client.get("setting/read.json")
+            account_id = settings["DEFAULT_EXCHANGE_DIFF_ACCOUNT_ID"]
+            return self._client.account_from_id(account_id)
+
+        def set_fx_gain_loss_account(account: int):
+            """Sets the FX gain/loss account in the settings."""
+            account_id = self._client.account_to_id(account)
+            payload = {"DEFAULT_EXCHANGE_DIFF_ACCOUNT_ID": account_id}
+            self._client.post("setting/update.json", params=payload)
+
+        # Get initial setting
+        initial_fx_gain_loss_account = get_fx_gain_loss_account()
+
+        if accounts is None:
+            ex_diff = self._client.get("fiscalperiod/exchangediff.json")["data"]
+            accounts = pd.DataFrame({
+                "foreign_currency_account": [
+                    self._client.account_from_id(entry['accountId']) for entry in ex_diff
+                ],
+                "fx_gain_loss_account": initial_fx_gain_loss_account,
+                "exchange_rate": None
+            })
+
+        # Record FX gain loss on given accounts
+        accounts = enforce_dtypes(accounts, FX_REVALUATION_ACCOUNT_COLUMNS)
```

Comment:
'exchange_rate' should be an optional column. `enforce_dtypes` automatically adds optional columns and fills them with NA.

Reply:
Btw: We can even increase flexibility and NA's in `fx_gain_loss_account` column. 
We can then replace NA values with `initial_fx_gain_loss_account`.